ragstack/colbertbase/colbertbase/astra_retriever.py

```diff
@@ -0,0 +1,140 @@
+
+    return max_sim
+'''
+
+# this torch based max similary has the best performance.
+# it is at least 20 times faster than dot product operator and numpy based implementation CuDA and CPU
+def max_similarity_torch(query_vector, embedding_list, is_cuda: bool=False):
+    """
+    Calculate the maximum similarity (dot product) between a query vector and a list of embedding vectors,
+    optimized for performance using PyTorch for GPU acceleration.
+
+    Parameters:
+    - query_vector: A PyTorch tensor representing the query vector.
+    - embedding_list: A list of PyTorch tensors, each representing an embedding vector.
+
+    Returns:
+    - max_sim: A float representing the highest similarity (dot product) score between the query vector
+               and the embedding vectors in the list, computed on the GPU.
+    """
+    # stacks the list of embedding tensors into a single tensor
+    if is_cuda:
+        query_vector = query_vector.to('cuda')
+        embedding_list = torch.stack(embedding_list).to('cuda')
+    else:
+        embedding_list = torch.stack(embedding_list)
+
+    # Calculate the dot products in a vectorized manner on the GPU
+    sims = torch.matmul(embedding_list, query_vector)
+
+    # Find the maximum similarity (dot product) value
+    max_sim = torch.max(sims)
+
+    # returns a tensor; the item() is the score
+    return max_sim
+
+
+class ColbertAstraRetriever():
+    astra: AstraDB
+    colbertEmbeddings: ColbertTokenEmbeddings
+    verbose: bool
+    is_cuda: bool=False
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def __init__(
+        self,
+        astraDB: AstraDB,
+        colbertEmbeddings: ColbertTokenEmbeddings,
+        verbose: bool=False,
+        **kwargs
+    ):
+        # initialize pydantic base model
+        self.astra = astraDB
+        self.colbertEmbeddings = colbertEmbeddings
+        self.verbose = verbose
+        self.is_cuda = torch.cuda.is_available()
+
+    def retrieve(
+        self,
+        query: str,
+        k: int=10,
+        query_maxlen: int=64,
+        **kwargs
+    ):
+        #
+        # if the query has fewer than a predefined number of of tokens Nq,
+        # colbertEmbeddings will pad it with BERT special [mast] token up to length Nq.
+        #
+        query_encodings = self.colbertEmbeddings.encode_query(query, query_maxlen=query_maxlen)
+
+        count = self.astra.session.execute(self.astra.chunk_counts_stmt).one().count
+        k = min(k, count)
+
+        # the min of query_maxlen is 32
+        top_k = max(math.floor(len(query_encodings) / 2), 16)
+        if self.verbose:
+            print(f"Total number of chunks: {count}, query length {len(query)} embeddings top_k: {top_k}")
+
+        # find the most relevant documents
+        docparts = set()
+        for qv in query_encodings:
+            # per token based retrieval
+            rows = self.astra.session.execute(self.astra.query_colbert_ann_stmt, [list(qv), top_k])
```

Comment:
perf gain to be had here with concurrent async queries…

Reply:
Good point! I changed to use futures and execute_async. Since Eq is much less than Ed, the improvement is still much better around 20% latency improvement